wizard.py

```diff
@@ -0,0 +1,1038 @@
+            ssc_inspections.append({
+              "name": inspect_name,
+              "cmd": run,
+              "based_on": step_name
+            })
+
+      ssc_steps.append(ssc_step)
+
+  ssc_data = {
+    "steps": ssc_steps,
+    "inspections": ssc_inspections
+  }
+  return ssc_data
+
+
+def form_data_to_ssc(step_names, step_commands, step_modifies,
+    inspection_names, inspection_commands, inspection_step_names):
+  """
+  <Purpose>
+    Takes step and inspection properties (form posted lists) and aggregates
+    them by index to step and inspection dictionaries to construct a dictionary
+    of software supply chain (ssc) data.
+
+  <Returns>
+    Software Supply Chain data, i.e. a dictionary of step and inspection data
+    Note: these are not not actual in-toto Step and Inspection objects
+
+    e.g.:
+    {
+      steps: [
+        {
+          "name": <unique step name>,
+          "cmd": <expected command>
+          "modifies": <boolean>  # whether the step modifies artifacts or not
+        }, ...
+      ],
+      inspections: [
+        {
+          "name": <unique inspection name>,
+          "cmd": <command to run inspecting>,
+          "based_on": <step name whose products are used for that inspection>
+        }
+      ]
+    }
+
+  """
+  steps = []
+  for i in range(len(step_names)):
+    steps.append({
+        "name": step_names[i],
+        "cmd": step_commands[i],
+        "modifies": step_modifies[i] == "true"
+      })
+
+  inspections = []
+  for i in range(len(inspection_names)):
+    inspections.append({
+        "name": inspection_names[i],
+        "cmd": inspection_commands[i],
+        "based_on": inspection_step_names[i]
+      })
+
+  ssc_data = {
+    "steps": steps,
+    "inspections": inspections
+  }
+  return ssc_data
+
+
+def _auth_items_to_dict(auth_items):
+  """Takes a list of auth_items and returns a dictionary mapping the items
+  to their respective step names, i.e..:
+  [{"step_name": ..., "threshold": ..., "authorized_functionaries": ...}, ...]
+  -->
+  {<step name> : {"step_name": ..., "threshold": ..., "authorized_..."}, ...}
+  """
+  auth_dict = {}
+  for auth_item in auth_items:
+    auth_dict[auth_item["step_name"]] = auth_item
+  return auth_dict
+
+
+# -----------------------------------------------------------------------------
+# NoSQL Helpers
+# -----------------------------------------------------------------------------
+# NOTE: Below functions rely on the current session having an id. If there is
+# no id in the session, all functions redirect to `404` (page not found).
+# This should never happen because all calling views should be decorated with
+# @with_session_id, which ensures that the current session does have an id.
+def _persist_session_subdocument(subdocument):
+  """Update a subdocument (e.g. vcs, ssc, functionaries...) in session document
+  identified by current session id. """
+  if not session.get("id"):
+    abort(404)
```

Comment:
I don't know if 404 is the error you want here. Could it be a 500 or a 501?

Reply:
404 made sense to me, because the user requests a resource that does not exist (without session). But 500 works as well.

Reply:
Hmm, I wonder if it's a privacy issue. We can continue with a 404 for the time being I imagine.